pkg/helpers/reconcile.go

```diff
@@ -0,0 +1,350 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var (
+	ErrInvalidSpec = fmt.Errorf("custom resource spec is invalid")
+	// ErrRequeue describes error while requeuing
+	ErrRequeue                    = fmt.Errorf("(transitory error) Requeue-ing resource to try again")
+	ErrWorkflowInFailureStatus    = errors.New("workflow in failure status")
+	ErrHelmReleaseInFailureStatus = errors.New("helmrelease in failure status")
+
+	dummyConfigmapYAMLSpec = `apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ .Release.Name }}-dummy
+  namespace: {{ .Release.Namespace }} 
+data:
+  name: {{ .Chart.Name }}
+  version : {{ .Chart.Version }}
+`
+	dummyConfigmapYAMLName = "templates/dummy-configmap.yaml"
+)
+
+type ReconcileHelper struct {
+	client.Client
+	logr.Logger
+	Instance       *v1alpha1.ApplicationGroup
+	Engine         workflow.Engine
+	RegistryClient *registry.Client
+
+	RegistryOptions RegistryClientOptions
+}
+
+type RegistryClientOptions struct {
+	StagingRepoName         string
+	TargetDir               string
+	CleanupDownloadedCharts bool
+}
+
+func (helper *ReconcileHelper) CreateOrUpdate(ctx context.Context) error {
+	helper.Logger = helper.Logger.WithValues(v1alpha1.AppGroupNameKey, helper.Instance.Name)
+	helper.V(3).Info("Reconciling ApplicationGroup object")
+
+	if len(helper.Instance.Spec.Applications) == 0 {
+		helper.Error(ErrInvalidSpec, "ApplicationGroup must list atleast one Application")
+		err := fmt.Errorf("application group must list atleast one Application : %w", ErrInvalidSpec)
+		return err
+	}
+
+	if err := helper.reconcileApplications(); err != nil {
+		helper.Error(err, "failed to reconcile the applications")
+		err = fmt.Errorf("failed to reconcile the applications : %w", err)
+		return err
+	}
+	// Generate the Workflow object to submit to Argo
+	if err := helper.generateWorkflow(ctx, helper.Logger, helper.Instance); err != nil {
+		helper.Error(err, "failed to reconcile ApplicationGroup instance")
+		return err
+	}
+	helper.Instance.Status.ObservedGeneration = helper.Instance.Generation
+	return nil
+}
+
+func (helper *ReconcileHelper) Rollback(ctx context.Context, patch client.Patch, err error) (ctrl.Result, error) {
+	// If this is a HelmRelease failure then we must remediate by cleaning up
+	// all the helm releases deployed by the workflow and helm operator
+	if errors.Is(err, ErrHelmReleaseInFailureStatus) {
+		// Delete the HelmRelease(s) - parent and subchart(s)
+		// Lookup charts using the label selector.
+		// Example: chart=kafka-dev,heritage=orkestra,owner=dev, where chart=<top-level-chart>
+		helper.Info("Remediating the applicationgroup with helmrelease failure status")
+		for _, app := range helper.Instance.Status.Applications {
+			switch meta.GetResourceCondition(&app.ChartStatus, meta.ReadyCondition).Reason {
+			case fluxhelmv2beta1.InstallFailedReason, fluxhelmv2beta1.UpgradeFailedReason, fluxhelmv2beta1.UninstallFailedReason,
```

Comment:
Move the resource conditions into a check function like IsFailedReason(<condition>) bool { // if condition in fluxhelmv2beta1.InstallFailedReason, fluxhelmv2beta1.UpgradeFailedReason, fluxhelmv2beta1.UninstallFailedReason, fluxhelmv2beta1.ArtifactFailedReason, fluxhelmv2beta1.InitFailedReason, fluxhelmv2beta1.GetLastReleaseFailedReason}